cpp/src/arrow/python/datetime.cc

```diff
@@ -262,6 +302,42 @@ int64_t PyDate_to_days(PyDateTime_Date* pydate) {
                             PyDateTime_GET_DAY(pydate));
 }
 
+// GIL must be held when calling this function.
+Status StringToTzinfo(const std::string& tz, PyObject** tzinfo) {
```

Comment:
It might go a bit too far (and can easily get outdated), but given that the below is basically python code written in C++, for me it would help in understanding it by including the short python snippet that it represents as a comment, eg

```
def string_to_tzinfo(tz):
    import pytz

    match = ... 
    if match:
        sign, hours, minutes = ...
        return pytz.FixedOffset(sign * (hours * 60 + minutes))
    else:
        return pytz.timezone(tz)
```

Reply:
Added a pointer back to this PR.  I providing the full code is overkill, and for those familiar with the CPP it should be fairly straight-forward to map back to python.  Let me know if you fell strongly about this.

Reply:
Looks fine!